src/policy/fees.cpp

```diff
@@ -76,116 +78,109 @@ void TxConfirmStats::UpdateMovingAverages()
-        // will be looking at the same amount of data and same bucket breaks)
+
+        // if we have no pending confirmations for this bucket we can continue, we do this because the decay rate
+        // can skew the data for a bucket making it seem like the bucket has a lower than 100%
+        // confirmation rate when in reality the bucket has had no pending transactions in it for a while
+        if (extraNum == 0)
+        {
+            continue;
+        }
+
+        // check for enough data points
         if (totalNum >= sufficientTxVal / (1 - decay))
         {
+            // find the rate at which transactions in this bucket are being confirmed
             double curPct = nConf / (totalNum + extraNum);
-
-            // Check to see if we are no longer getting confirmed at the success rate
-            if (requireGreater && curPct < successBreakPoint)
-                break;
-            if (!requireGreater && curPct > successBreakPoint)
-                break;
-
-            // Otherwise update the cumulative stats, and the bucket variables
-            // and reset the counters
-            else
+            if (curPct < successBreakPoint)
             {
-                foundAnswer = true;
-                nConf = 0;
-                totalNum = 0;
-                extraNum = 0;
-                bestNearBucket = curNearBucket;
-                bestFarBucket = curFarBucket;
-                curNearBucket = bucket + step;
+                selectedBucket = bucket;
+                break;
             }
+            nConf = 0;
+            totalNum = 0;
+            extraNum = 0;
         }
     }
+    // if our confirm rate for any bucket is never less than 80% selectedBucket will
+    // be -1 at the end of the loop.
+    // so we return mintxfee
+    if (selectedBucket < 0)
+    {
+        return minTxFee;
+    }
 
     double median = -1;
-    double txSum = 0;
-
-    // Calculate the "average" fee of the best bucket range that met success conditions
-    // Find the bucket with the median transaction and then report the average fee from that bucket
-    // This is a compromise between finding the median which we can't since we don't save all tx's
-    // and reporting the average which is less accurate
-    unsigned int minBucket = bestNearBucket < bestFarBucket ? bestNearBucket : bestFarBucket;
-    unsigned int maxBucket = bestNearBucket > bestFarBucket ? bestNearBucket : bestFarBucket;
-    for (unsigned int j = minBucket; j <= maxBucket; j++)
+
+    // check if the historical moving average of txs in this bucket is 0
+    if (txCtAvg[selectedBucket] != 0)
     {
-        txSum += txCtAvg[j];
+        // if it is not, we are in the right bucket
+        median = avg[selectedBucket] / txCtAvg[selectedBucket];
     }
-    if (foundAnswer && txSum != 0)
+    // if we didnt error but somehow got a value less than the mintxfee return the mintxfee
+    if (median > 0 && median < minTxFee)
```

Comment:
If the median == -1 (i.e. txCtAvg[selectedBucket] == 0) shouldn't we just default to the minTxFee as well?

Reply:
good catch, i missed this edge case. this shouldnt realistically ever happen but it is good to not return an error when it does

Reply:
looks like -1 will still be returned if txCtAvg[selectedBucket]==0

Reply:
oh, i misread your comment. i fixed the other issue with selecting a negative array index, fixed this now as well.